public/main/inc/lib/formvalidator/Rule/Username.php

```diff
@@ -21,10 +21,12 @@ class HTML_QuickForm_Rule_Username extends HTML_QuickForm_Rule
      */
     public function validate($username, $options)
     {
-        if ('true' == api_get_setting('login_is_email')) {
+        // Allow field username to use "@" and "." BT#19778
+        return api_valid_email($username);
```

Comment:
Hola @ywarnier , yo tenia esa duda  al comienzo, debido a que ya existe esa validacion cuando el username es un correo login_is_email, como iba a ser aplicado para ambos casos mande el PR de esa forma, ahora analizando el codigo anterior veo que usa purify_username($username, true) , donde es validado con URLify::transliterate , "transliterate" transforma el simbolo @ en "(at)" y no valida el username, lo cual no cumple el pedido de la tarea, la otra forma seria decirle que no sea estricto purify_username($username, false) , es este caso solo valida "Applying the shallow purifier" (USERNAME_PURIFIER_SHALLOW) y "Length limitation" (USERNAME_MAX_LENGTH) , seria lo correcto para esta version de chamilo ? , URLify es una libreria que esta en el vendor. Otra solucion que veo es cambiar el parametro de USERNAME_PURIFIER_SHALLOW por USERNAME_PURIFIER agregando el @ con purify_username($username, false)

Gracias por la confirmacion.

Reply:
USERNAME_PURIFIER_SHALLOW *solo* elimina los espacios en blanco. Es un poco demasiado ligero, igual con USERNAME_MAX_LENGTH.
El URLify::transliterate(), a cambio, me parece hacer un paso demasiado transformando el @ en `(at)`. No lo sabía.
Iba a decir que el problema de usar USERNAME_PURIFIER en vez de USERNAME_PURIFIER_SHALLOW es que no permite caracteres fuera de ASCII, pero el transliterate() de todas formas tampoco.
Bueno entonces simplemente
1. agrega @ dentro de USERNAME_PURIFIER
2. agrega una conversión inversa de `(at)` a `@` después del transliterate(). Creo que es lo más sencillo, lo que modifica menos el comportamiento actual y no debería tener mayor impacto otro. Pero ponle un comentario tipo
```
// We want everything transliterate() does except converting @ to '(at)'. This is a hack to avoid this.
$return = str_replace('(at)', '@', $return);
```

Reply:
Gracias @ywarnier , corregido en el siguiente commit